index.js

```diff
@@ -701,6 +723,74 @@
     );
   };
 
+  //  test :: [Type] -> Type -> Any -> Boolean
+  var test = function(env) {
+    return function recur(t) {
+      return function(x) {
+        switch (t.type) {
+          case 'ANY':
+            return true;
+          case 'UNKNOWN':
+            return true;
+          case 'VARIABLE':
+            return true;
+          case 'NULLARY':
+            return t.test(x);
+          case 'UNARY':
+            var xs = t._1(x);
+            return t.name === 'sanctuary-def/StrMap' ?
+              ($.Object.test(x) &&
+               (isEmpty(xs) ||
+                !isEmpty(consistentTypes(env)(xs)) && all(xs, recur(t.$1)))) :
+              t.test(x) && all(xs, recur(t.$1));
```

Comment:
This is pretty awful, but I don't think it can be simplified. To be a member of `sanctuary-def/StrMap`, a value must:

  - satisfy `$.Object.test`; and
  - either:
      - be empty; or
      - have values which are all members of at least one type and which a members of `t.$1`.

Actually, it seems the last point could be separated into two: one for type variables; one for other types.